<a href="https://colab.research.google.com/github/ameesha543/Statistical-Learning-e23095/blob/main/Assignment_5/Answers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Foundations of Statistical Inference & Hypothesis Testing**

Objective of this assignment is to rigorously evaluate the stationarity of an operational baseline and master the mathematical machinery required to control false-alarm risks in continuous diagnostic layers.

## **Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)**



1. Consider the sequence of i.i.d. random variables  
$$\mathbf{X}_1,\mathbf{X}_2,\cdots, \mathbf{X}_n\in \mathbb{R}^m$$
with $\boldsymbol{\mu}=\mathbb{E}(\mathbf{X}_i)$ and $\boldsymbol{\Sigma}=\mathrm{Var}(\mathbf{X}_i)$. The unobservable multivariate population parameters $(\boldsymbol{\mu}, \boldsymbol{\Sigma})$ are estimated with their corresponding unbiased empirical sample estimators $(\widehat{\boldsymbol{\mu}}_n, \mathbf{S})$, where
$$\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_{i},$$
and
$$\mathbf{S} = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T .$$

* Formally derive why the raw Maximum Likelihood Estimator for the covariance matrix, $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$, is inherently **biased** in finite samples, by showing that:

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n-1}{n}\boldsymbol{\Sigma}$$


* Mathematically demonstrate how applying Bessel's correction yields the unique, unbiased estimator $\mathbf{S}$.


2. Under the null hypothesis $H_0$ that a structural asset remains completely healthy (the status quo), a chosen test statistic tracks a deterministic parametric distribution.
* Define a **Type I Error ($\alpha$)** and a **Type II Error ($\beta$)** explicitly within the context of continuous structural health monitoring.
* If an engineer sets an ultra-conservative significance threshold (e.g., $\alpha = 0.0001$) to prevent false alarms from shutting down an industrial plant, explain the mathematical consequence this decision exerts on the statistical power of the diagnostic layer. How does this modify the geometric volume of your "healthy operation" confidence ellipsoid?

# 1. Bias of the Maximum Likelihood Estimator (MLE) of the Covariance Matrix

Let

$$
\widehat{\boldsymbol{\mu}}_n
=
\frac{1}{n}
\sum_{i=1}^{n}
\mathbf{X}_i
$$

and define the Maximum Likelihood Estimator (MLE) of the covariance matrix as

$$
\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}
=
\frac{1}{n}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T.
$$

To determine whether this estimator is unbiased, consider the identity

$$
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
=
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
-
n
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T.
$$

Taking expectations on both sides,

$$
\mathbb{E}
\left[
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
\right]
=
\mathbb{E}
\left[
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
\right]
-
n
\mathbb{E}
\left[
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T
\right].
$$

Since

$$
\mathbb{E}
\left[
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
\right]
=
\boldsymbol{\Sigma},
$$

we obtain

$$
\mathbb{E}
\left[
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
\right]
=
n\boldsymbol{\Sigma}.
$$

Also,

$$
\mathrm{Var}
(\widehat{\boldsymbol{\mu}}_n)
=
\frac{\boldsymbol{\Sigma}}{n},
$$

which implies

$$
\mathbb{E}
\left[
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T
\right]
=
\frac{\boldsymbol{\Sigma}}{n}.
$$

Substituting,

$$
\mathbb{E}
\left[
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
\right]
=
n\boldsymbol{\Sigma}
-
n
\left(
\frac{\boldsymbol{\Sigma}}{n}
\right)
=
(n-1)\boldsymbol{\Sigma}.
$$

Therefore,

$$
\mathbb{E}
[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}]
=
\frac{1}{n}
(n-1)\boldsymbol{\Sigma}
=
\frac{n-1}{n}
\boldsymbol{\Sigma}.
$$

Hence,

$$
\boxed{
\mathbb{E}
[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}]
=
\frac{n-1}{n}
\boldsymbol{\Sigma}
}
$$

which proves that the raw MLE underestimates the true covariance matrix and is therefore biased.

---

## Bessel's Correction and the Unbiased Estimator

To eliminate this bias, we replace the denominator \(n\) with \(n-1\). The corrected estimator is

$$
\mathbf{S}
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T.
$$

Taking expectations,

$$
\mathbb{E}[\mathbf{S}]
=
\frac{1}{n-1}
\mathbb{E}
\left[
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
\right].
$$

Using the result derived above,

$$
\mathbb{E}[\mathbf{S}]
=
\frac{1}{n-1}
(n-1)\boldsymbol{\Sigma}
=
\boldsymbol{\Sigma}.
$$

Therefore,

$$
\boxed{
\mathbb{E}[\mathbf{S}]
=
\boldsymbol{\Sigma}
}
$$

showing that \(\mathbf{S}\) is an unbiased estimator of the population covariance matrix.

The correction is called **Bessel's correction** and compensates for the loss of one degree of freedom caused by estimating the population mean from the same sample.

---

# 2. Type I and Type II Errors in Structural Health Monitoring

Assume the null hypothesis is

$$
H_0:\text{The structure is healthy}.
$$

and the alternative hypothesis is

$$
H_1:\text{The structure is damaged}.
$$

## Type I Error (\(\alpha\))

A Type I Error occurs when the monitoring system rejects \(H_0\) even though the structure is actually healthy.

Mathematically,

$$
\boxed{
\alpha
=
P(\text{Reject } H_0 \mid H_0 \text{ is true})
}
$$

In structural health monitoring, this corresponds to a **false alarm**, where a healthy structure is incorrectly classified as damaged. Such errors may trigger unnecessary inspections, maintenance procedures, or costly plant shutdowns.

---

## Type II Error (\(\beta\))

A Type II Error occurs when the monitoring system fails to reject \(H_0\) even though damage is actually present.

Mathematically,

$$
\boxed{
\beta
=
P(\text{Fail to reject } H_0 \mid H_1 \text{ is true})
}
$$

In structural health monitoring, this corresponds to **missed damage detection**, where an unhealthy structure is incorrectly classified as healthy.

---

## Effect of an Ultra-Conservative Significance Level

Suppose the engineer chooses

$$
\alpha = 0.0001.
$$

The statistical power of the test is

$$
\boxed{
\text{Power}
=
1-\beta
}
$$

which measures the probability of correctly detecting actual damage.

Reducing \(\alpha\) makes the rejection region much smaller because stronger evidence is required before damage is declared.

Consequently,

$$
\alpha \downarrow
\quad \Longrightarrow \quad
\beta \uparrow
\quad \Longrightarrow \quad
(1-\beta) \downarrow.
$$

Thus, lowering the significance level decreases the probability of false alarms but also reduces the probability of detecting real damage.

---

## Effect on the Healthy-Operation Confidence Ellipsoid

In multivariate monitoring systems, healthy observations are often represented by a confidence ellipsoid defined by

$$
(\mathbf{x}-\boldsymbol{\mu})^T
\boldsymbol{\Sigma}^{-1}
(\mathbf{x}-\boldsymbol{\mu})
\leq c_{\alpha},
$$

where \(c_{\alpha}\) is a critical threshold chosen so that

$$
P
\left(
(\mathbf{X}-\boldsymbol{\mu})^T
\boldsymbol{\Sigma}^{-1}
(\mathbf{X}-\boldsymbol{\mu})
\leq c_{\alpha}
\right)
=
1-\alpha.
$$

When \(\alpha\) is reduced,

$$
1-\alpha
\rightarrow 1,
$$

which increases the critical value \(c_{\alpha}\).

Therefore,

$$
\boxed{
\alpha \downarrow
\quad \Longrightarrow \quad
c_{\alpha} \uparrow
}
$$

and the healthy-operation ellipsoid expands.

Since the volume of the ellipsoid is proportional to

$$
V
\propto
(c_{\alpha})^{m/2},
$$

where \(m\) is the dimension of the feature space,

$$
\boxed{
V
\propto
(c_{\alpha})^{m/2}
}
$$

a larger critical value produces a larger confidence ellipsoid.

Therefore, choosing

$$
\alpha = 0.0001
$$

results in:

- A much larger healthy-operation region.
- Fewer false alarms.
- Lower statistical power.
- Greater risk of missing actual structural damage.

Thus, an ultra-conservative significance threshold improves operational stability but reduces damage-detection sensitivity.

## **Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)**

The Lindeberg-Lévy Multivariate Central Limit Theorem establishes that as the sample snapshot size approaches infinity ($n \to \infty$), the standardized difference between the empirical mean and population mean converges in distribution to a Multivariate Gaussian:

$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$



However, the true population covariance matrix $\boldsymbol{\Sigma}$ is a latent, unobservable parameter.
* State **Slutsky’s Theorem** rigorously.
* Prove how Slutsky's Theorem justifies substituting the empirical sample covariance matrix $\mathbf{S}$ for the latent population matrix $\boldsymbol{\Sigma}$ as $n \to \infty$, thereby operationalizing the practical parametric modeling distribution:

$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left({\boldsymbol{\mu}}, \frac{1}{n}\mathbf{S}\right)$$







# Slutsky's Theorem and the Practical Distribution of the Sample Mean

The Multivariate Central Limit Theorem states that for a sequence of i.i.d. random vectors

$$
\mathbf{X}_1,\mathbf{X}_2,\ldots,\mathbf{X}_n \in \mathbb{R}^m,
$$

with population mean

$$
\boldsymbol{\mu}
=
\mathbb{E}[\mathbf{X}_i]
$$

and covariance matrix

$$
\boldsymbol{\Sigma}
=
\mathrm{Var}(\mathbf{X}_i),
$$

the sample mean

$$
\widehat{\boldsymbol{\mu}}_n
=
\frac{1}{n}
\sum_{i=1}^{n}
\mathbf{X}_i
$$

satisfies

$$
\sqrt{n}
\left(
\widehat{\boldsymbol{\mu}}_n
-
\boldsymbol{\mu}
\right)
\xrightarrow{d}
\mathscr{N}
(\mathbf{0},\boldsymbol{\Sigma}).
$$

This result depends on the unknown population covariance matrix
\(\boldsymbol{\Sigma}\), which is generally not observable in practice.

---

# Slutsky's Theorem

Let

$$
X_n \xrightarrow{d} X
$$

and

$$
Y_n \xrightarrow{p} c,
$$

where \(c\) is a constant.

Then Slutsky's Theorem states that

$$
X_n + Y_n
\xrightarrow{d}
X + c,
$$

$$
X_nY_n
\xrightarrow{d}
cX,
$$

and, provided \(c\neq 0\),

$$
\frac{X_n}{Y_n}
\xrightarrow{d}
\frac{X}{c}.
$$

More generally, if a sequence of random matrices converges in probability to a fixed matrix, then that matrix may be replaced by its consistent estimator without affecting the asymptotic distribution.

---

# Consistency of the Sample Covariance Matrix

The sample covariance matrix is

$$
\mathbf{S}
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T.
$$

Under standard regularity conditions,

$$
\mathbf{S}
\xrightarrow{p}
\boldsymbol{\Sigma}.
$$

This means that the sample covariance matrix is a consistent estimator of the true population covariance matrix.

Therefore, as the sample size increases,

$$
\mathbf{S}
-
\boldsymbol{\Sigma}
\xrightarrow{p}
\mathbf{0}.
$$

---

# Applying Slutsky's Theorem

From the Multivariate Central Limit Theorem,

$$
\sqrt{n}
\left(
\widehat{\boldsymbol{\mu}}_n
-
\boldsymbol{\mu}
\right)
\xrightarrow{d}
\mathscr{N}
(\mathbf{0},\boldsymbol{\Sigma}).
$$

At the same time,

$$
\mathbf{S}
\xrightarrow{p}
\boldsymbol{\Sigma}.
$$

Since the covariance estimator converges in probability to the true covariance matrix, Slutsky's Theorem allows us to replace the unknown matrix \(\boldsymbol{\Sigma}\) by its consistent estimator \(\mathbf{S}\) without changing the asymptotic distribution.

Consequently,

$$
\sqrt{n}
\left(
\widehat{\boldsymbol{\mu}}_n
-
\boldsymbol{\mu}
\right)
\xrightarrow{d}
\mathscr{N}
(\mathbf{0},\mathbf{S}).
$$

Equivalently,

$$
\widehat{\boldsymbol{\mu}}_n
\xrightarrow{d}
\mathscr{N}
\left(
\boldsymbol{\mu},
\frac{1}{n}\mathbf{S}
\right).
$$

Thus, for sufficiently large sample sizes,

$$
\boxed{
\widehat{\boldsymbol{\mu}}_n
\sim
\mathscr{N}
\left(
\boldsymbol{\mu},
\frac{1}{n}\mathbf{S}
\right)
}
$$

which is the practical parametric model used in statistical inference.

---

# Interpretation

The Multivariate Central Limit Theorem provides the asymptotic Gaussian form of the sample mean but depends on the unknown covariance matrix \(\boldsymbol{\Sigma}\).

Because the sample covariance matrix \(\mathbf{S}\) converges in probability to \(\boldsymbol{\Sigma}\), Slutsky's Theorem guarantees that replacing the unknown population covariance by its empirical estimate does not alter the asymptotic distribution.

Therefore, in practical engineering and data-analysis applications, the sample mean can be modeled as

$$
\widehat{\boldsymbol{\mu}}_n
\sim
\mathscr{N}
\left(
\boldsymbol{\mu},
\frac{1}{n}\mathbf{S}
\right),
$$

allowing confidence regions, hypothesis tests, and uncertainty quantification to be constructed using only observable sample statistics.

## **Part C: Numerical Verification**

***Scenario:** You are provided an industrial dataset containing $n = 5,000$ snapshots across $m = 3$ sensor channels measuring physical strain. To verify the weak i.i.d. assumption before initializing a global PCA or Factor Analysis layer, you must explicitly test **Condition 1: Constant Joint Mean (First Moment Homogeneity)** across the time index.*

**Your Task:**
Write a clean, self-contained Python function within Google Colab to partition the dataset into $g = 5$ consecutive, non-overlapping blocks (chunks) of size $n_j = 1,000$. Implement a Multivariate Analysis of Variance (MANOVA) tracking framework from scratch (or leveraging `statsmodels`) to calculate **Wilks' Lambda ($\Lambda$)**.

$$\Lambda = \frac{|\mathbf{W}|}{|\mathbf{B} + \mathbf{W}|}$$

Transform your calculated $\Lambda$ using **Bartlett’s $\chi^2$ approximation**:


$$\chi^2_{\text{calc}} = -\left(n - 1 - \frac{m + g}{2}\right) \ln \Lambda$$

Execute your code using the synthetic data generator provided below. Report your exact calculated Wilks' $\Lambda$, the Bartlett $\chi^2$ test statistic, and its corresponding $p$-value under $\chi^2 \sim \chi^2_{m(g-1)}$. Based on an alpha threshold of $\alpha = 0.05$, state your final diagnostic conclusion: Has the baseline center shifted over time, or is the first moment homogeneous?



```python
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

# Generating base stationary multivariate normal noise
base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g chunks and evaluates first-moment homogeneity
    via Wilks' Lambda and Bartlett's Chi-Square asymptotic transformation.
    """
    # YOUR CODE GOES HERE
    # 1. Split df into g equal consecutive blocks
    # 2. Compute the global mean and individual chunk means
    # 3. Calculate Within-Chunk Matrix (W) and Between-Chunk Matrix (B)
    # 4. Compute Wilks' Lambda, Bartlett's Chi-Square, and the final p-value
    pass

# Execute your completed function below:
# results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
# print(results)
```

# **Geometric Subspace Optimization via Principal Component Analysis (PCA)**

The objective of this assignment is to master the rigorous linear algebra mechanics underpinning coordinate transformations, verify total variance conservation laws, and design operational monitoring thresholds using the Hotelling's $T^2$ and $Q$ (SPE) statistics.

## **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

1. Let $\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^m$ represent a zero-mean, centered baseline feature vector with a symmetric, positive-semidefinite population covariance matrix $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$. We apply the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$, and define the orthogonal projection mapping:

$$\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$$


* Prove step-by-step that the theoretical covariance matrix of the transformed vector, $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$, simplifies exactly to the strictly diagonal matrix $\mathbf{\Lambda}$.
* Explain the precise geometric and statistical meaning of this diagonal result regarding the cross-covariances between distinct coordinates $Z_{i,j}$ and $Z_{i,k}$ (for $j \neq k$).


2. The total variance of a multivariate system can be measured in its raw coordinate space or its transformed principal component space.
* Utilizing the algebraic properties of the matrix trace ($\text{tr}$), specifically its invariance under cyclic permutations, formally prove that the total system variance is preserved during the mapping from raw space to the de-correlated principal space:

$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$


* If we partition the empirical matrix into an explained subspace ($\widehat{\mathbf{P}}_k \in \mathbb{R}^{m \times k}$) and an unexplained residual subspace ($\widehat{\mathbf{P}}_{m-k} \in \mathbb{R}^{m \times (m-k)}$), state the rigorous mathematical formulations for the **Cumulative Explained Variance Ratio $\Phi(k)$** and the **Residual Unexplained Variance Ratio $\Psi(k)$**.


3. In structural health monitoring (SHM), we truncate our coordinate system to retain only the first $k$ components ($k < m$). The centered raw vector is partitioned as $(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) = \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$.
* Define the reconstructed vector $\widehat{\mathbf{x}}_i$ and the residual error vector $\mathbf{e}_i$.
* Prove via the Pythagorean identity that the total squared Euclidean norm of the centered observation vector cleanly decomposes into independent subspace energies:

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$


* Contrast the diagnostic functions of Hotelling's $T^2$ statistic and the residual $Q$ statistic (SPE). If a structure encounters an environmental load anomaly (e.g., extreme operational wind loads) versus an internal structural fracture (e.g., a fatigue crack changing the structural load pathways), match each event type to the statistic ($T^2$ or $Q$) most sensitive to it and justify your choices.

# 1. Covariance Structure in Principal Component Space

Let the centered random vector be

$$
\widetilde{\mathbf{X}}_i
=
\mathbf{X}_i-\boldsymbol{\mu}
$$

with covariance matrix

$$
\boldsymbol{\Sigma}
=
\mathbb{E}
\left[
\widetilde{\mathbf{X}}_i
\widetilde{\mathbf{X}}_i^T
\right].
$$

Since \(\boldsymbol{\Sigma}\) is symmetric and positive semidefinite, it admits the spectral decomposition

$$
\boldsymbol{\Sigma}
=
\mathbf{P}
\mathbf{\Lambda}
\mathbf{P}^T,
$$

where

$$
\mathbf{P}
=
[\mathbf{p}_1,\mathbf{p}_2,\ldots,\mathbf{p}_m]
$$

is an orthogonal matrix satisfying

$$
\mathbf{P}^T\mathbf{P}
=
\mathbf{I},
$$

and

$$
\mathbf{\Lambda}
=
\mathrm{diag}
(\lambda_1,\lambda_2,\ldots,\lambda_m).
$$

Define the principal component transformation

$$
\mathbf{Z}_i
=
\mathbf{P}^T
\widetilde{\mathbf{X}}_i.
$$

---

## Covariance of the Transformed Vector

The covariance matrix of \(\mathbf{Z}_i\) is

$$
\mathrm{Cov}(\mathbf{Z}_i)
=
\mathbb{E}
[\mathbf{Z}_i\mathbf{Z}_i^T].
$$

Substituting the transformation,

$$
=
\mathbb{E}
\left[
(\mathbf{P}^T\widetilde{\mathbf{X}}_i)
(\mathbf{P}^T\widetilde{\mathbf{X}}_i)^T
\right].
$$

Using

$$
(\mathbf{P}^T\widetilde{\mathbf{X}}_i)^T
=
\widetilde{\mathbf{X}}_i^T\mathbf{P},
$$

we obtain

$$
=
\mathbf{P}^T
\mathbb{E}
\left[
\widetilde{\mathbf{X}}_i
\widetilde{\mathbf{X}}_i^T
\right]
\mathbf{P}.
$$

Since

$$
\mathbb{E}
\left[
\widetilde{\mathbf{X}}_i
\widetilde{\mathbf{X}}_i^T
\right]
=
\boldsymbol{\Sigma},
$$

we have

$$
=
\mathbf{P}^T
\boldsymbol{\Sigma}
\mathbf{P}.
$$

Substituting the spectral decomposition,

$$
=
\mathbf{P}^T
(\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T)
\mathbf{P}.
$$

Using orthogonality,

$$
\mathbf{P}^T\mathbf{P}
=
\mathbf{I},
$$

gives

$$
=
\mathbf{\Lambda}.
$$

Hence,

$$
\boxed{
\mathbb{E}
[\mathbf{Z}_i\mathbf{Z}_i^T]
=
\mathbf{\Lambda}
}
$$

which is a diagonal matrix.

---

## Geometric and Statistical Interpretation

Since

$$
\mathbf{\Lambda}
=
\mathrm{diag}
(\lambda_1,\lambda_2,\ldots,\lambda_m),
$$

all off-diagonal entries are zero.

Therefore,

$$
\boxed{
\mathrm{Cov}(Z_{i,j},Z_{i,k})
=
0,
\qquad j\neq k
}
$$

meaning that distinct principal components are uncorrelated.

Geometrically:

- PCA rotates the coordinate system.
- The new axes coincide with the directions of maximum variance.
- The transformed coordinates are mutually orthogonal.

Statistically:

- Each principal component captures an independent source of variation.
- No linear redundancy remains between different coordinates.

---

# 2. Preservation of Total Variance

The total variance of the system is defined as

$$
\mathrm{tr}
(\boldsymbol{\Sigma}).
$$

Using the spectral decomposition,

$$
\mathrm{tr}
(\boldsymbol{\Sigma})
=
\mathrm{tr}
(\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T).
$$

Using cyclic invariance of the trace,

$$
\mathrm{tr}(ABC)
=
\mathrm{tr}(BCA),
$$

we obtain

$$
=
\mathrm{tr}
(\mathbf{\Lambda}\mathbf{P}^T\mathbf{P}).
$$

Since

$$
\mathbf{P}^T\mathbf{P}
=
\mathbf{I},
$$

it follows that

$$
=
\mathrm{tr}
(\mathbf{\Lambda}).
$$

Because \(\mathbf{\Lambda}\) is diagonal,

$$
\mathrm{tr}
(\mathbf{\Lambda})
=
\sum_{j=1}^{m}
\lambda_j.
$$

Therefore,

$$
\boxed{
\mathrm{tr}
(\boldsymbol{\Sigma})
=
\mathrm{tr}
(\mathbf{\Lambda})
=
\sum_{j=1}^{m}
\lambda_j
}
$$

which proves that PCA preserves total variance.

---

## Cumulative Explained Variance Ratio

If only the first \(k\) principal components are retained,

$$
\boxed{
\Phi(k)
=
\frac{\sum_{j=1}^{k}\lambda_j}
{\sum_{j=1}^{m}\lambda_j}
}
$$

This quantity measures the proportion of total variance explained by the retained subspace.

---

## Residual Unexplained Variance Ratio

The unexplained variance ratio is

$$
\boxed{
\Psi(k)
=
\frac{\sum_{j=k+1}^{m}\lambda_j}
{\sum_{j=1}^{m}\lambda_j}
}
$$

Equivalently,

$$
\boxed{
\Psi(k)
=
1-\Phi(k)
}
$$

which measures the variance discarded by dimensionality reduction.

---

# 3. Reconstruction and Residual Analysis

The centered observation can be decomposed as

$$
(\mathbf{x}_i-\widehat{\boldsymbol{\mu}}_n)
=
\widehat{\mathbf{P}}_k\mathbf{z}_{i,k}
+
\widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}.
$$

---

## Reconstructed Vector

The PCA reconstruction using only the first \(k\) components is

$$
\boxed{
\widehat{\mathbf{x}}_i
=
\widehat{\boldsymbol{\mu}}_n
+
\widehat{\mathbf{P}}_k
\mathbf{z}_{i,k}
}
$$

---

## Residual Error Vector

The residual vector is

$$
\mathbf{e}_i
=
\mathbf{x}_i
-
\widehat{\mathbf{x}}_i.
$$

Substituting the reconstruction,

$$
\boxed{
\mathbf{e}_i
=
\widehat{\mathbf{P}}_{m-k}
\mathbf{z}_{i,m-k}
}
$$

which lies entirely in the discarded subspace.

---

## Pythagorean Energy Decomposition

Since

$$
\widehat{\mathbf{P}}_k
\perp
\widehat{\mathbf{P}}_{m-k},
$$

the retained and residual subspaces are orthogonal.

Hence,

$$
\left\|
\mathbf{x}_i-\widehat{\boldsymbol{\mu}}_n
\right\|^2
=
\left\|
\widehat{\mathbf{P}}_k
\mathbf{z}_{i,k}
+
\widehat{\mathbf{P}}_{m-k}
\mathbf{z}_{i,m-k}
\right\|^2.
$$

Expanding,

$$
=
\|\widehat{\mathbf{P}}_k\mathbf{z}_{i,k}\|^2
+
\|\widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}\|^2
+
2
(\widehat{\mathbf{P}}_k\mathbf{z}_{i,k})^T
(\widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}).
$$

The cross-term vanishes because of orthogonality:

$$
(\widehat{\mathbf{P}}_k)^T
\widehat{\mathbf{P}}_{m-k}
=
\mathbf{0}.
$$

Therefore,

$$
=
\|\mathbf{z}_{i,k}\|^2
+
\|\mathbf{z}_{i,m-k}\|^2.
$$

Hence,

$$
\boxed{
\|\mathbf{x}_i-\widehat{\boldsymbol{\mu}}_n\|^2
=
\|\mathbf{z}_{i,k}\|^2
+
\|\mathbf{z}_{i,m-k}\|^2
}
$$

which is the PCA Pythagorean identity.

---

# Hotelling's \(T^2\) Statistic vs Q Statistic (SPE)

## Hotelling's \(T^2\)

The \(T^2\) statistic measures variation within the retained PCA subspace:

$$
\boxed{
T^2
=
\mathbf{z}_{i,k}^T
\mathbf{\Lambda}_k^{-1}
\mathbf{z}_{i,k}
}
$$

It detects abnormal behavior that still follows the established correlation structure.

Examples:

- Extreme wind loads
- Heavy operational loading
- Temperature-driven operating changes

These events increase the magnitude of scores in the principal subspace.

Therefore,

$$
\boxed{
\text{Environmental / operational anomalies}
\rightarrow T^2
}
$$

---

## Q Statistic (Squared Prediction Error)

The Q statistic measures residual energy outside the retained PCA model:

$$
\boxed{
Q
=
\|\mathbf{e}_i\|^2
=
\|\mathbf{z}_{i,m-k}\|^2
}
$$

It detects new behavior not represented by the baseline covariance structure.

Examples:

- Fatigue cracks
- Structural fractures
- Sensor failures
- Changes in load transfer pathways

Such events create energy in directions excluded from the normal PCA model.

Therefore,

$$
\boxed{
\text{Structural damage or fatigue crack}
\rightarrow Q
}
$$

---

## Summary

$$
\boxed{
\text{Extreme operational load anomaly}
\rightarrow T^2
}
$$

because the event remains inside the learned PCA subspace.

$$
\boxed{
\text{Internal structural fracture}
\rightarrow Q
}
$$

because the damage introduces new variance directions outside the baseline model.

## **Part B: Numerical Verification**



***Scenario:** You are building a real-time diagnostics layer for a structural component monitored by $m = 4$ sensors. You have a baseline archive of $n = 3,000$ nominal snapshots. To establish operational boundaries, you must write an optimization pipeline that analyzes how Hotelling’s $T^2$ and the residual $Q$ statistic respond to different truncation boundaries $k \in \{1, 2, 3\}$.*

**Your Task:**
Complete the Python class method below inside Google Colab. The function must:

1. Standardize and center the incoming realizations.
2. Decompose the unbiased sample covariance matrix $\mathbf{S}$ using `np.linalg.eigh` and sort components in descending order ($\widehat{\lambda}_1 \ge \dots \ge \widehat{\lambda}_m$).
3. Loop through every possible subspace cutoff value $k \in \{1, 2, 3\}$, calculating the sample vector array for Hotelling’s $T^2$ ($T^2_i = \sum_{j=1}^k \frac{z_{i,j}^2}{\widehat{\lambda}_j}$) and the $Q$ statistic ($Q_i = \sum_{j=k+1}^m z_{i,j}^2$).
4. Compute the expected value (empirical mean) of $T^2$ and $Q$ across all samples for each $k$.
5. Return these metrics to verify your theoretical derivations in Part A.



In [ ]:
import numpy as np
import pandas as pd

class SHMPCAAnalyzer:

    def evaluate_subspace_statistics(self, X):
        """
        Evaluate Hotelling's T² and Q statistics for
        k = 1, 2, ..., m-1 principal components.

        Parameters
        ----------
        X : ndarray of shape (n_samples, n_features)

        Returns
        -------
        results : pd.DataFrame
        """

        # --------------------------------------------------
        # 1. Center the data
        # --------------------------------------------------
        X_centered = X - np.mean(X, axis=0)

        n, m = X_centered.shape

        # --------------------------------------------------
        # 2. Unbiased covariance matrix
        # --------------------------------------------------
        S = np.cov(X_centered, rowvar=False, bias=False)

        # --------------------------------------------------
        # 3. Eigen decomposition
        # --------------------------------------------------
        eigvals, eigvecs = np.linalg.eigh(S)

        # Sort in descending order
        idx = np.argsort(eigvals)[::-1]

        eigvals = eigvals[idx]
        eigvecs = eigvecs[:, idx]

        # --------------------------------------------------
        # PCA scores
        # --------------------------------------------------
        Z = X_centered @ eigvecs

        results = []

        # --------------------------------------------------
        # 4. Loop over k = 1,2,...,m-1
        # --------------------------------------------------
        for k in range(1, m):

            # -------- Hotelling T² --------
            T2 = np.sum(
                (Z[:, :k] ** 2) / eigvals[:k],
                axis=1
            )

            # -------- Q Statistic (SPE) --------
            Q = np.sum(
                Z[:, k:] ** 2,
                axis=1
            )

            results.append({
                "k": k,
                "Mean_T2": np.mean(T2),
                "Mean_Q": np.mean(Q)
            })

        return pd.DataFrame(results)

In [ ]:
# Synthetic baseline archive
np.random.seed(42)

X = np.random.multivariate_normal(
    mean=[0,0,0,0],
    cov=[[1.0,0.5,0.3,0.2],
         [0.5,1.0,0.4,0.1],
         [0.3,0.4,1.0,0.6],
         [0.2,0.1,0.6,1.0]],
    size=3000
)

analyzer = SHMPCAAnalyzer()

results = analyzer.evaluate_subspace_statistics(X)

print(results)

   k   Mean_T2    Mean_Q
0  1  0.999667  1.931969
1  2  1.999333  0.862293
2  3  2.999000  0.311624


# **Latent Subspace Decomposition via Factor Analysis (FA)**

The objective is to master the generative model equations of Factor Analysis, derive factor scores from scratch using conditional expectations (Thomson's method), and numerically isolate sensor malfunctions from true structural damage by tracking communalities ($h^2$) and uniquenesses ($\varphi^2$).

## **Part A: Theoretical Fundamentals (Generative Model & Commonalities)**

1. Consider your structured data setup where $\mathbf{Z}_i \in \mathbb{R}^m$ is a normalized, zero-mean baseline feature vector whose population covariance matrix represents the universal correlation matrix $\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$. The Factor Analysis model posits the linear generative structure:
$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$
where $\mathbf{F}_i \in \mathbb{R}^k$ are the orthogonal common factors ($\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] = \mathbf{I}_{k \times k}$) and $\boldsymbol{\epsilon}_i \in \mathbb{R}^m$ is the diagonalized measurement noise matrix ($\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \dots, \varphi_m^2)$).
* Utilizing the constraint that the latent factors and errors are independent ($\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}$), prove the **Fundamental Equation of Factor Analysis**:

$$\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$


* Focus on the $j$-th diagonal entry of this decomposition. Explicitly define the **Communality ($h_j^2$)** and **Uniqueness ($\varphi_j^2$)** parameters, and explain their physical interpretations in an asset diagnostics system.


2. In Factor Analysis, we often apply an orthogonal coordinate rotation to the factor loading matrix (such as the **Varimax** rotation criterion) to improve model interpretability.
* Contrast how a raw principal component loading vector differs conceptually from a Varimax-rotated factor loading vector.
* Explain why orthogonal rotations can alter the individual columns of the loading matrix $\boldsymbol{\Lambda}$ without changing the calculated communality ($h_j^2$) of any sensor or modifying the global covariance approximation $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$.



1. Because the common factors $\mathbf{F}_i$ are unobservable latent parameters, they cannot be directly extracted via geometric projections like PCA. Instead, we must treat them as hidden random vectors and calculate their realization using **Thomson’s Regression Method**, which relies on the conditional expectation $\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{z}_i]$.
* Construct the joint partitioned vector $\mathbf{Y}_i = [\mathbf{Z}_i^T, \mathbf{F}_i^T]^T$ and prove that its joint population covariance matrix expands exactly to:

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix}$$


* Under the standard assumption of a Multivariate Normal distribution, the conditional mean of a partitioned system is governed by the projection identity $\mathbb{E}[\mathbf{X}_2 | \mathbf{x}_1] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{x}_1 - \boldsymbol{\mu}_1)$. Using this theorem, derive the final score matrix equation for Thomson’s estimator:

$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i$$


# 1. Derivation of Thomson's Regression Factor Score Estimator

Consider the Factor Analysis model

$$
\mathbf{Z}_i
=
\boldsymbol{\Lambda}\mathbf{F}_i
+
\boldsymbol{\epsilon}_i
$$

where

$$
\mathbb{E}[\mathbf{F}_i]
=
\mathbf{0},
\qquad
\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]
=
\mathbf{I}_{k\times k},
$$

and

$$
\mathbb{E}[\boldsymbol{\epsilon}_i]
=
\mathbf{0},
\qquad
\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]
=
\boldsymbol{\Psi}.
$$

The common factors and measurement errors are assumed independent:

$$
\mathbb{E}
[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]
=
\mathbf{0}.
$$

The correlation matrix of the observed variables is

$$
\mathbf{R}
=
\boldsymbol{\Lambda}
\boldsymbol{\Lambda}^T
+
\boldsymbol{\Psi}.
$$

---

## Construction of the Joint Vector

Define the partitioned random vector

$$
\mathbf{Y}_i
=
\begin{bmatrix}
\mathbf{Z}_i \\
\mathbf{F}_i
\end{bmatrix}.
$$

Since both vectors are centered,

$$
\mathbb{E}
[\mathbf{Y}_i]
=
\mathbf{0}.
$$

The covariance matrix of the joint vector is

$$
\boldsymbol{\Sigma}_{YY}
=
\mathrm{Cov}
(\mathbf{Y}_i)
=
\mathbb{E}
\left[
\mathbf{Y}_i\mathbf{Y}_i^T
\right].
$$

Expanding the block covariance structure,

$$
\boldsymbol{\Sigma}_{YY}
=
\begin{bmatrix}
\mathrm{Cov}(\mathbf{Z}_i)
&
\mathrm{Cov}(\mathbf{Z}_i,\mathbf{F}_i)
\\
\mathrm{Cov}(\mathbf{F}_i,\mathbf{Z}_i)
&
\mathrm{Cov}(\mathbf{F}_i)
\end{bmatrix}.
$$

---

### Upper-Left Block

The covariance matrix of the observed variables is

$$
\mathrm{Cov}(\mathbf{Z}_i)
=
\mathbf{R}.
$$

---

### Upper-Right Block

Using the factor model,

$$
\mathrm{Cov}
(\mathbf{Z}_i,\mathbf{F}_i)
=
\mathbb{E}
\left[
(\boldsymbol{\Lambda}\mathbf{F}_i+\boldsymbol{\epsilon}_i)
\mathbf{F}_i^T
\right].
$$

Expanding,

$$
=
\boldsymbol{\Lambda}
\mathbb{E}
[\mathbf{F}_i\mathbf{F}_i^T]
+
\mathbb{E}
[\boldsymbol{\epsilon}_i\mathbf{F}_i^T].
$$

Since

$$
\mathbb{E}
[\mathbf{F}_i\mathbf{F}_i^T]
=
\mathbf{I},
$$

and

$$
\mathbb{E}
[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]
=
\mathbf{0},
$$

it follows that

$$
\mathrm{Cov}
(\mathbf{Z}_i,\mathbf{F}_i)
=
\boldsymbol{\Lambda}.
$$

---

### Lower-Left Block

Using symmetry of covariance matrices,

$$
\mathrm{Cov}
(\mathbf{F}_i,\mathbf{Z}_i)
=
\boldsymbol{\Lambda}^T.
$$

---

### Lower-Right Block

Since the factors are standardized,

$$
\mathrm{Cov}
(\mathbf{F}_i)
=
\mathbf{I}_{k\times k}.
$$

---

Combining all four blocks,

$$
\boxed{
\boldsymbol{\Sigma}_{YY}
=
\begin{bmatrix}
\mathbf{R}
&
\boldsymbol{\Lambda}
\\
\boldsymbol{\Lambda}^T
&
\mathbf{I}_{k\times k}
\end{bmatrix}
}
$$

which proves the required result.

---

# Derivation of Thomson's Regression Estimator

Assume the joint vector

$$
\mathbf{Y}_i
=
\begin{bmatrix}
\mathbf{Z}_i \\
\mathbf{F}_i
\end{bmatrix}
$$

follows a multivariate normal distribution.

For a partitioned multivariate normal vector

$$
\begin{bmatrix}
\mathbf{X}_1 \\
\mathbf{X}_2
\end{bmatrix},
$$

the conditional expectation is

$$
\mathbb{E}
[\mathbf{X}_2|\mathbf{x}_1]
=
\boldsymbol{\mu}_2
+
\boldsymbol{\Sigma}_{21}
\boldsymbol{\Sigma}_{11}^{-1}
(\mathbf{x}_1-\boldsymbol{\mu}_1).
$$

For the factor model, identify

$$
\mathbf{X}_1=\mathbf{Z}_i,
\qquad
\mathbf{X}_2=\mathbf{F}_i.
$$

Since all variables are centered,

$$
\boldsymbol{\mu}_1
=
\mathbf{0},
\qquad
\boldsymbol{\mu}_2
=
\mathbf{0}.
$$

Therefore,

$$
\mathbb{E}
[\mathbf{F}_i|\mathbf{z}_i]
=
\boldsymbol{\Sigma}_{21}
\boldsymbol{\Sigma}_{11}^{-1}
\mathbf{z}_i.
$$

From the previously derived joint covariance matrix,

$$
\boldsymbol{\Sigma}_{21}
=
\boldsymbol{\Lambda}^T,
$$

and

$$
\boldsymbol{\Sigma}_{11}
=
\mathbf{R}.
$$

Substituting these blocks into the conditional mean formula gives

$$
\mathbb{E}
[\mathbf{F}_i|\mathbf{z}_i]
=
\boldsymbol{\Lambda}^T
\mathbf{R}^{-1}
\mathbf{z}_i.
$$

Thomson's regression method defines the estimated factor score as

$$
\mathbf{f}_i
=
\mathbb{E}
[\mathbf{F}_i|\mathbf{z}_i].
$$

Hence,

$$
\boxed{
\mathbf{f}_i
=
\boldsymbol{\Lambda}^T
\mathbf{R}^{-1}
\mathbf{z}_i
}
$$

which is the Thomson Regression Factor Score Estimator.

---

## Interpretation

The matrix

$$
\boldsymbol{\Lambda}^T
\mathbf{R}^{-1}
$$

acts as a linear projection operator that transforms the observed standardized measurements into estimates of the latent factor values.

Therefore, Thomson's estimator provides the minimum mean-square error linear prediction of the unobservable common factors based on the observed data.

## **Part B: Parametric Factor Partition Engine & UI Architecture**

The objective is to implement an end-to-end data processing and diagnostic visualization pipeline that instantiates a parametric **Factor Analysis (FA)** engine with **Varimax rotation**. The engine must partition empirical observation metrics into a shared structural subspace and a distinct diagonal noise floor, mapping the results onto a highly structured, asymmetric grid interface.

### Develop a Mathematical Engine & Pre-processing Guardrails


Implement an isolated algorithm to structure and decompose empirical data structures according to the following guidelines:

1. **Z-Score Mapping Layer:** Accept an arbitrary data frame and isolate numeric observation columns. Standardize the observations into standard normal space $Z \in \mathbb{R}^{n \times m}$ applying an unbiased sample standard deviation ($ddof=1$). Build an explicit numerical guardrail that catches zero-variance channels ($\sigma_j = 0$) and caps them to a minimum boundary threshold of $10^{-15}$ to prevent divide-by-zero runtime exceptions.
2. **Dimensionality Constraint Layer:** Let $m$ represent the count of isolated sensor features, and $k$ represent the target count of latent physical factors. Program two validation rule exceptions:
* Reject processing if the operational snapshot count ($n$) is less than or equal to the channel count ($m$).
* Throw an descriptive error if the requested latent space dimension ($k$) is not strictly less than the available feature space dimension ($m$).


3. **Information Decomposition Layer:** Fit a Factor Analysis framework utilizing a Maximum Likelihood or Singular Value Decomposition (SVD) algorithm combined with an orthogonal Varimax rotation matrix ($T$). From the fitted parameters, compute the vector arrays for:
* **Communality Vector ($h^2 \in \mathbb{R}^m$):** The total shared variance captured across each channel.
* **Uniqueness Vector ($\varphi^2 \in \mathbb{R}^m$):** The remaining isolated sensor noise floor.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis


def varimax(Phi, gamma=1.0, q=50, tol=1e-6):
    """
    Orthogonal Varimax rotation.
    """
    p, k = Phi.shape

    R = np.eye(k)
    d_old = 0

    for _ in range(q):

        Lambda = Phi @ R

        U, S, Vt = np.linalg.svd(
            Phi.T @ (
                Lambda**3
                - (gamma / p)
                * Lambda
                @ np.diag(np.sum(Lambda**2, axis=0))
            )
        )

        R = U @ Vt

        d = np.sum(S)

        if d_old != 0 and (d - d_old) < tol:
            break

        d_old = d

    return Phi @ R, R


def factor_analysis_pipeline(df, k):
    """
    Factor Analysis pipeline with:
        1. Z-score standardization
        2. Dimension validation
        3. Factor extraction
        4. Varimax rotation
        5. Communality & uniqueness computation
    """

    # --------------------------------------------------
    # 1. Isolate numeric columns
    # --------------------------------------------------

    X = df.select_dtypes(include=[np.number])

    if X.empty:
        raise ValueError(
            "No numeric observation columns found."
        )

    n, m = X.shape

    # --------------------------------------------------
    # 2. Dimension validation
    # --------------------------------------------------

    if n <= m:
        raise ValueError(
            f"Invalid dimensions: n={n}, m={m}. "
            "Requirement: n > m."
        )

    if k >= m:
        raise ValueError(
            f"Invalid latent dimension k={k}. "
            "Requirement: k < m."
        )

    # --------------------------------------------------
    # 3. Z-score mapping
    # --------------------------------------------------

    mu = X.mean(axis=0)

    sigma = X.std(axis=0, ddof=1)

    sigma = np.maximum(sigma, 1e-15)

    Z = (X - mu) / sigma

    Z = Z.to_numpy()

    # --------------------------------------------------
    # 4. Factor Analysis (Maximum Likelihood)
    # --------------------------------------------------

    fa = FactorAnalysis(
        n_components=k,
        random_state=42
    )

    fa.fit(Z)

    loadings = fa.components_.T

    # --------------------------------------------------
    # 5. Varimax rotation
    # --------------------------------------------------

    rotated_loadings, rotation_matrix = varimax(loadings)

    # --------------------------------------------------
    # 6. Communality vector
    # --------------------------------------------------

    communalities = np.sum(
        rotated_loadings**2,
        axis=1
    )

    # --------------------------------------------------
    # 7. Uniqueness vector
    # --------------------------------------------------

    uniqueness = 1.0 - communalities

    uniqueness = np.maximum(
        uniqueness,
        0.0
    )

    # --------------------------------------------------
    # Results
    # --------------------------------------------------

    results = {
        "Z": pd.DataFrame(
            Z,
            columns=X.columns
        ),

        "Rotated_Loadings": pd.DataFrame(
            rotated_loadings,
            index=X.columns,
            columns=[
                f"Factor_{i+1}"
                for i in range(k)
            ]
        ),

        "Communality_h2": pd.Series(
            communalities,
            index=X.columns
        ),

        "Uniqueness_phi2": pd.Series(
            uniqueness,
            index=X.columns
        ),

        "Rotation_Matrix": rotation_matrix
    }

    return results

In [ ]:
# Example dataset

np.random.seed(42)

df = pd.DataFrame(
    np.random.randn(3000, 5),
    columns=[
        "Sensor_1",
        "Sensor_2",
        "Sensor_3",
        "Sensor_4",
        "Sensor_5"
    ]
)

results = factor_analysis_pipeline(df, k=2)

print("Communality Vector (h²)")
print(results["Communality_h2"])

print("\nUniqueness Vector (φ²)")
print(results["Uniqueness_phi2"])

print("\nRotated Factor Loadings")
print(results["Rotated_Loadings"])

Communality Vector (h²)
Sensor_1    0.020073
Sensor_2    0.055816
Sensor_3    0.006087
Sensor_4    0.014692
Sensor_5    0.032481
dtype: float64

Uniqueness Vector (φ²)
Sensor_1    0.979927
Sensor_2    0.944184
Sensor_3    0.993913
Sensor_4    0.985308
Sensor_5    0.967519
dtype: float64

Rotated Factor Loadings
          Factor_1  Factor_2
Sensor_1  0.094068  0.105947
Sensor_2 -0.235059  0.023738
Sensor_3  0.073942 -0.024882
Sensor_4  0.106181  0.058460
Sensor_5  0.071627 -0.165379


### Latent Score Projection

1. **Mathematical State Reconstruction:** Using Thomson’s Minimum Mean Squared Error (MMSE) regression methodology, transform the standard normal snapshot array ($Z$) directly into hidden factor states $F \in \mathbb{R}^{n \times k}$ using the system relation:

$$F = Z R^{-1} \Lambda_{\text{rotated}}$$


2. **Performance Diagnostic Telemetry:** Calculate the empirical variance of the reconstructed factor scores enforcing an unbiased scale correction ($ddof=1$). Print a clean confirmation breakdown out to the logging console that presents the calculated system-wide metrics:

$$\text{Average System Communality \%} = \frac{1}{m}\sum_{j=1}^m h_j^2 \times 100$$


$$\text{Average System Uniqueness \%} = \frac{1}{m}\sum_{j=1}^m \varphi_j^2 \times 100$$

In [ ]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis

# ==================================================
# 1. Dataset Generation & Preprocessing
# ==================================================
np.random.seed(42)

df = pd.DataFrame(
    np.random.randn(3000, 6),
    columns=["S1", "S2", "S3", "S4", "S5", "S6"]
)

# Factor analysis operates on standardized data (Z-scores)
# This is required for the correlation-based Thomson MMSE scoring below
Z = (df - df.mean()) / df.std(ddof=0)
Z = Z.to_numpy()

# ==================================================
# 2. Factor Analysis Model Execution
# ==================================================
k = 2
fa = FactorAnalysis(n_components=k, rotation="varimax", random_state=42)
fa.fit(Z)

# Extract parameters needed for the diagnostic pipeline
rotated_loadings = fa.components_.T
uniqueness = fa.noise_variance_
communalities = np.sum(rotated_loadings**2, axis=1)

# ==================================================
# 3. Correlation Matrix & Numerical Stabilization
# ==================================================
R = np.corrcoef(Z, rowvar=False)

# Numerical stabilization
R_inv = np.linalg.pinv(R)

# ==================================================
# 4. Thomson MMSE Factor Score Estimation
#
# F = Z R^{-1} Lambda_rotated
# ==================================================
F = Z @ R_inv @ rotated_loadings

# ==================================================
# 5. Empirical Factor Variances
# (unbiased estimator)
# ==================================================
factor_variances = np.var(
    F,
    axis=0,
    ddof=1
)

# ==================================================
# 6. System Performance Metrics
# ==================================================
avg_communality_pct = (
    np.mean(communalities) * 100
)

avg_uniqueness_pct = (
    np.mean(uniqueness) * 100
)

# ==================================================
# 7. Diagnostic Console Output
# ==================================================
print("=" * 60)
print("FACTOR ANALYSIS DIAGNOSTIC SUMMARY")
print("=" * 60)

print(
    f"Average System Communality (%) : "
    f"{avg_communality_pct:.3f}"
)

print(
    f"Average System Uniqueness (%)  : "
    f"{avg_uniqueness_pct:.3f}"
)

print("\nFactor Score Variances:")

for idx, var in enumerate(factor_variances, start=1):
    print(
        f"Factor {idx}: {var:.6f}"
    )

print("=" * 60)

FACTOR ANALYSIS DIAGNOSTIC SUMMARY
Average System Communality (%) : 3.443
Average System Uniqueness (%)  : 96.593

Factor Score Variances:
Factor 1: 0.099372
Factor 2: 0.098802


### Layout Engineering & Interface Design

Render a unified, production-grade **2 × 2 diagnostic canvas** using Plotly Subplots that enforces the following strict geometric layout constraints to ensure maximum readability and prevent overlapping label blocks:

* **Grid Specifications:** Configure the dashboard matrix with an explicit `horizontal_spacing=0.24` and a `vertical_spacing=0.28`. Set the global canvas parameters to `width=1250` and `height=750` utilizing a clean `plotly_white` theme.
* **Legend Formatting:** Place a single, horizontally oriented legend centered directly above the subplots (`orientation="h", x=0.5, y=1.02`). Enforce clean `margin` padding structures to isolate the interface elements (`t=150, b=60, l=140, r=80`).

Populate the grid layout using the four specific tracking profiles below:

```
                  2 × 2 FACTOR DIAGNOSTICS GRID MATRIX
+------------------------------------------+------------------------------------------+
|             [Row 1, Col 1]               |             [Row 1, Col 2]               |
|      Structural Loadings Heatmap         |     Horizontal Stacked Bar Chart         |
|   - Absolute values |λ_(j,r)|            |   - Shared Communality (h²) in Blue      |
|   - Color scale: 'YlOrRd'                |   - Isolated Uniqueness (φ²) in Orange   |
|   - Left-anchored scale bar (x = -0.15)  |   - Total space constraints: 0 to 100%   |
+------------------------------------------+------------------------------------------+
|             [Row 2, Col 1]               |             [Row 2, Col 2]               |
|         Sensor Noise Floor Line          |       Latent Empirical Variance          |
|   - Dot-dashed red line graph with markers|   - Vertical distribution bar chart      |
|   - Tracks exact Uniqueness values (φ²)  |   - Verifies energy level allocations    |
|   - Enforce tick angle adjustments (25°) |   - Black boundary outlines (width = 0.5)|
+------------------------------------------+------------------------------------------+

```

1. **Subplot (1,1): Structural Loadings Matrix Heatmap**
Render the absolute value of the rotated component weights matrix ($|\Lambda_{\text{rotated}}|$) using a continuous `YlOrRd` (Yellow-Orange-Red) color palette. To protect the clarity of the sensor text identifiers along the y-axis, isolate the continuous scale colorbar to the left side of the canvas by explicitly setting `x=-0.15`, `len=0.38`, `y=0.78`, `yanchor="middle"`, and `xanchor="right"`.
2. **Subplot (1,2): Variance Partitioning Profile**
Build a horizontal stacked bar chart mapping out the tracking allocation per sensor channel. The base layer must represent the percentage contribution of shared communalities ($h^2 \times 100$) utilizing color hex code `#1f77b4`. Stack the corresponding sensor uniqueness values ($\varphi^2 \times 100$) directly on top utilizing color hex code `#ff7f0e`. Force the x-axis limits strictly between `[0, 100]`.
3. **Subplot (2,1): Sensor Uniqueness Line Profile**
Construct a clear line chart tracking the isolated channel noise thresholds ($\varphi^2$) across all monitored hardware positions. Apply a red (`#d62728`), 2-pixel wide dot-dashed line string combined with visual `x` indicators. Apply a custom horizontal label layout optimization by rotating the category ticks to a distinct `tickangle=25` constraint.
4. **Subplot (2,2): Latent Factor Variance Distribution**
Draw a vertical bar chart displaying the empirical variance computed across the extracted common factor score vectors. Color the bars green (`#2ca02c`) with an explicit black border envelope (`width=0.5`).

## Thomson MMSE State Reconstruction and Performance Diagnostic Telemetry

The standardized observation matrix \(Z\) is transformed into latent factor scores using Thomson's Minimum Mean Squared Error (MMSE) regression estimator:

$$
F = ZR^{-1}\Lambda_{\text{rotated}}
$$

where:

- \(Z \in \mathbb{R}^{n \times m}\) is the standardized observation matrix,
- \(R\) is the correlation matrix of the standardized observations,
- \(R^{-1}\) is the inverse (or pseudo-inverse) of the correlation matrix,
- \(\Lambda_{\text{rotated}}\) is the Varimax-rotated factor loading matrix,
- \(F \in \mathbb{R}^{n \times k}\) contains the estimated latent factor states.

This estimator provides the minimum mean squared error linear prediction of the hidden factors based on the observed measurements.

---

### Empirical Factor Variance

The empirical variance of each reconstructed factor is computed using the unbiased estimator:

$$
\widehat{\mathrm{Var}}(F_j)
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(F_{ij}-\bar{F}_j)^2
\qquad (j=1,\ldots,k)
$$

which corresponds to using `ddof=1` in NumPy.

---

### Average System Communality

The communality of sensor \(j\) is defined as

$$
h_j^2
=
\sum_{r=1}^{k}
\lambda_{jr}^2
$$

where \(\lambda_{jr}\) denotes the rotated loading of sensor \(j\) on factor \(r\).

The average system communality percentage is

$$
\boxed{
\text{Average System Communality (\%)}
=
\frac{1}{m}
\sum_{j=1}^{m}
h_j^2
\times 100
}
$$

This metric represents the average proportion of sensor variance explained by the common latent factors.

---

### Average System Uniqueness

The uniqueness of sensor \(j\) is defined as

$$
\varphi_j^2
=
1-h_j^2
$$

The average system uniqueness percentage is

$$
\boxed{
\text{Average System Uniqueness (\%)}
=
\frac{1}{m}
\sum_{j=1}^{m}
\varphi_j^2
\times 100
}
$$

This metric represents the average proportion of variance attributed to sensor-specific noise, local disturbances, and unexplained effects.

---

### Interpretation

A high average communality indicates that the sensor network is strongly governed by common latent physical mechanisms and that the factor model explains a large fraction of the observed variability.

A high average uniqueness indicates that a substantial portion of the observed variance remains sensor-specific and is not captured by the latent factor structure.

Since

$$
h_j^2+\varphi_j^2=1,
$$

the average communality percentage and average uniqueness percentage should approximately sum to \(100\%\), apart from minor numerical rounding effects.

# **Subspace Diagnostics & Feature De-correlation**

In structural health monitoring (SHM) and industrial asset diagnostics, multi-sensor grids generate highly continuous, collinear data streams. To establish reliable anomaly detection limits, engineers must project these high-dimensional observation spaces onto lower-dimensional sub-spaces.

This assignment evaluates two fundamental methods for dimensionality reduction using an identical 4-channel sensor array (`Sensor_1` to `Sensor_4`):

1. **Principal Component Analysis (PCA):** A geometric, non-parametric approach aimed at total variance maximization.
2. **Factor Analysis (FA):** A generative, parametric framework aimed at partitioning shared structural variance from unique localized sensor noise.



## **Part A: Develop a Feature Analysis Dashboard**

**UI Layout 1: PCA Optimization Suite ($2\times  3$ Horizontal Layout)**

The first interface must display a 5-panel horizontal grid layout.
* **Panel 1,1 (Heatmap):** Feature Loadings Matrix $|P_{j,r}|$.
* **Panel 1,2 (Bar Chart):** Absolute Eigenvalues ($\lambda_j$) tracking variance magnitude.
* **Panel 1,3 (Combined Bar + Line):** Marginal Explained Variance % per axis (bars) overlaid with Cumulative Captured Variance Curve (dashed line).
* **Panel 2,1 (Bar Chart):** Residual Space (Unexplained Variance %) showing excluded information ($100 - \text{Cumulative Variance \%}$).
* **Panel 2,2 (Scatter Line):** Sample Mean Hotelling's $T^2$ tracking distance vs. Subspace Size $k$.
* **Panel 2,3 (Scatter Line):** Sample Mean $Q$ Statistic (SPE) tracking residual decay vs. Truncation Cutoff $k$.

**UI Layout 2: FA Latent Subspace Suite ($2 \times 2$ Grid Layout)**

The second interface must structure its subplots into a balanced $2 \times 2$ matrix.

* **Panel 1,1 (Heatmap):** Structural Loadings Matrix $|\lambda_{j,r}|$.
* **Panel 1,2 (Stacked Horizontal Bar):** Variance Allocation breakdown showing Shared Communality ($h^2$) versus Channel Uniqueness ($\varphi^2$) totaling $100\%$ per sensor.
* **Panel 2,1 (Scatter Line):** Sensor Uniqueness Noise Floor Profile ($\varphi^2$) across monitored channels to instantly isolate instrument faults.
* **Panel 2,2 (Bar Chart):** Latent Factor Scores Empirical Variance showing the post-rotation power of the extracted physical dimensions.


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

def plot_pca_suite(loadings,
                   eigenvalues,
                   explained_variance_ratio,
                   mean_t2,
                   mean_q):

    cumulative_variance = np.cumsum(explained_variance_ratio) * 100
    explained_pct = explained_variance_ratio * 100
    residual_pct = 100 - cumulative_variance

    fig, axes = plt.subplots(
        2, 3,
        figsize=(18, 10)
    )

    # --------------------------------------------------
    # Panel 1,1 : Loading Heatmap
    # --------------------------------------------------
    ax = axes[0,0]

    im = ax.imshow(
        np.abs(loadings),
        aspect='auto'
    )

    ax.set_title("Feature Loadings |P(j,r)|")
    ax.set_xlabel("Principal Components")
    ax.set_ylabel("Sensors")

    plt.colorbar(im, ax=ax)

    # --------------------------------------------------
    # Panel 1,2 : Eigenvalues
    # --------------------------------------------------
    ax = axes[0,1]

    ax.bar(
        np.arange(1, len(eigenvalues)+1),
        eigenvalues
    )

    ax.set_title("Eigenvalue Magnitudes")
    ax.set_xlabel("Component")
    ax.set_ylabel("Eigenvalue")

    # --------------------------------------------------
    # Panel 1,3 : Explained Variance
    # --------------------------------------------------
    ax = axes[0,2]

    x = np.arange(1, len(eigenvalues)+1)

    ax.bar(
        x,
        explained_pct,
        label="Explained Variance (%)"
    )

    ax.plot(
        x,
        cumulative_variance,
        '--o',
        linewidth=2,
        label="Cumulative Variance (%)"
    )

    ax.set_title("Variance Capture")
    ax.legend()

    # --------------------------------------------------
    # Panel 2,1 : Residual Variance
    # --------------------------------------------------
    ax = axes[1,0]

    ax.bar(
        x,
        residual_pct
    )

    ax.set_title("Residual Variance (%)")
    ax.set_xlabel("k")
    ax.set_ylabel("%")

    # --------------------------------------------------
    # Panel 2,2 : Mean T²
    # --------------------------------------------------
    ax = axes[1,1]

    ax.plot(
        np.arange(1, len(mean_t2)+1),
        mean_t2,
        'o-'
    )

    ax.set_title("Mean Hotelling T²")
    ax.set_xlabel("Subspace Size k")

    # --------------------------------------------------
    # Panel 2,3 : Mean Q
    # --------------------------------------------------
    ax = axes[1,2]

    ax.plot(
        np.arange(1, len(mean_q)+1),
        mean_q,
        'o-'
    )

    ax.set_title("Mean Q Statistic (SPE)")
    ax.set_xlabel("Subspace Size k")

    plt.tight_layout()
    plt.show()

In [ ]:
def plot_fa_suite(loadings,
                  communalities,
                  uniqueness,
                  factor_variances):

    fig, axes = plt.subplots(
        2, 2,
        figsize=(14, 10)
    )

    sensor_names = [
        f"Sensor {i+1}"
        for i in range(len(communalities))
    ]

    # --------------------------------------------------
    # Panel 1,1 : Loading Heatmap
    # --------------------------------------------------
    ax = axes[0,0]

    im = ax.imshow(
        np.abs(loadings),
        aspect='auto'
    )

    ax.set_title("Structural Loadings |λ(j,r)|")
    ax.set_xlabel("Factors")
    ax.set_ylabel("Sensors")

    plt.colorbar(im, ax=ax)

    # --------------------------------------------------
    # Panel 1,2 : Communality vs Uniqueness
    # --------------------------------------------------
    ax = axes[0,1]

    ax.barh(
        sensor_names,
        communalities,
        label="Communality"
    )

    ax.barh(
        sensor_names,
        uniqueness,
        left=communalities,
        label="Uniqueness"
    )

    ax.set_title("Variance Allocation")
    ax.legend()

    # --------------------------------------------------
    # Panel 2,1 : Uniqueness Profile
    # --------------------------------------------------
    ax = axes[1,0]

    ax.plot(
        sensor_names,
        uniqueness,
        'o-'
    )

    ax.set_title("Sensor Uniqueness Profile")
    ax.set_ylabel("φ²")

    # --------------------------------------------------
    # Panel 2,2 : Factor Variance
    # --------------------------------------------------
    ax = axes[1,1]

    ax.bar(
        np.arange(1, len(factor_variances)+1),
        factor_variances
    )

    ax.set_title("Factor Score Variances")
    ax.set_xlabel("Factor")

    plt.tight_layout()
    plt.show()

## **Part B: Analysis**

### Generate the following synthetic data and answer the questions below

The synthetic multi-sensor asset data must be generated using
```python
import numpy as np
import pandas as pd

# Set execution seed for perfect reproducible properties
np.random.seed(42)
n_samples = 2500

# 1. Generate underlying, independent true physical driving forces (Latent Factors)
f1 = np.random.normal(0, 1, n_samples)  # Primary structural mode/load
f2 = np.random.normal(0, 1, n_samples)  # Secondary operational mode

# 2. Construct the Observable Multi-Sensor Array with targeted noise profiles
# Sensors 1 and 2 couple strongly to the primary process (f1)
s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)

# Sensor 3 couples strongly to the secondary process (f2)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)

# Sensor 4 suffers an extreme electrical/instrumentation fault (massive localized white noise)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

# 3. Compile into the final evaluation DataFrame
df_asset = pd.DataFrame(
    data=np.vstack([s1, s2, s3, s4]).T,
    columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
)

```

**Empirical Covariance and Correlation Properties**

* **Sensors 1, 2, and 3:** High cross-correlation due to strong coupling with the common underlying physical structures ($f_1$ and $f_2$).
* **Sensor 4:** Highly un-correlated with all other channels, but exhibits an exceptionally large variance magnitude ($\sigma^2 \approx 2.0$) due to its high noise contamination scale.


### Question 1: The Total Variance Illusion (PCA Subplots 1 & 2 vs. FA Subplots 1 & 2)

* **Context:** In your PCA dashboard, `PC 1` captures roughly $45\%$ of the total system variance, and `PC 2` captures over $30\%$, indicating a highly structured, multi-component system. However, looking at the Factor Analysis dashboard, the **Variance Partitioning Stacked Bar** reveals that `Sensor_4` has a **Uniqueness value ($\varphi^2$) exceeding $98\%$**.
* **Core Problem:** 1. What does a Uniqueness value close to $100\%$ tell you about the physical nature of the data coming from `Sensor_4`?
2. Because PCA maximizes total global variance without distinguishing between types of variance, how does the massive localized noise of `Sensor_4` "trick" the PCA engine into inflating its top eigenvalues?
3. Why is this a major risk if an engineer relies *only* on PCA to build an automated plant anomaly detection framework?



## Interpretation of the PCA and Factor Analysis Results

### 1. Physical Meaning of a Uniqueness Value Near 100%

The uniqueness of a sensor is defined as

$$
\varphi_j^2 = 1 - h_j^2,
$$

where \(h_j^2\) is the communality of the sensor.

If `Sensor_4` exhibits a uniqueness value exceeding \(98\%\),

$$
\varphi_4^2 > 0.98,
$$

then less than \(2\%\) of its variance is explained by the common latent factors governing the system.

This indicates that the measurements from `Sensor_4` are largely independent of the underlying structural behavior captured by the other sensors. Most of its variability originates from sensor-specific effects such as:

- Measurement noise,
- Electrical interference,
- Calibration errors,
- Local environmental disturbances,
- Instrument degradation or malfunction.

Therefore, `Sensor_4` contributes very little information about the shared physical state of the monitored asset and behaves almost like an isolated noise source.

---

### 2. How Sensor_4 Can Inflate PCA Eigenvalues

Principal Component Analysis (PCA) seeks directions that maximize total variance.

Mathematically, PCA decomposes the covariance matrix

$$
\Sigma = P \Lambda P^T,
$$

where the eigenvalues

$$
\lambda_1,\lambda_2,\ldots,\lambda_m
$$

represent the amount of variance captured along each principal direction.

A key limitation of PCA is that it does not distinguish between:

- Useful structural variance,
- Sensor noise variance.

Both contribute equally to the covariance matrix.

Since `Sensor_4` contains an unusually large amount of variance that is not shared with the rest of the sensor network, PCA interprets this variance as important information rather than noise.

Consequently:

- The covariance matrix gains additional variance from `Sensor_4`.
- One or more eigenvalues become artificially larger.
- PCA allocates principal components to explain the noise generated by `Sensor_4`.

In effect, the localized noise of `Sensor_4` "tricks" PCA into believing that the noise represents a meaningful source of system variation.

---

### 3. Risk of Using Only PCA for Anomaly Detection

This behavior creates a significant risk in automated structural health monitoring and industrial diagnostics.

Because PCA focuses exclusively on maximizing variance:

- No distinction is made between genuine structural behavior and sensor-specific noise.
- Principal components may become partially dominated by faulty sensors.
- Detection thresholds based on Hotelling's \(T^2\) may be distorted.
- The retained PCA subspace may represent noise rather than physical system dynamics.

As a result:

1. **False Alarms (Type I Errors):**
   Random fluctuations from `Sensor_4` may be interpreted as abnormal system behavior.

2. **Missed Damage Detection (Type II Errors):**
   Genuine structural changes may be masked because part of the PCA model capacity is wasted explaining sensor noise.

3. **Unstable Monitoring Models:**
   The principal components become sensitive to sensor faults rather than true changes in structural condition.

Factor Analysis helps overcome this limitation because it explicitly separates:

$$
\text{Total Variance}
=
\text{Shared Structural Variance}
+
\text{Unique Sensor Variance}.
$$

The communality term identifies variance associated with common physical mechanisms, while the uniqueness term isolates sensor-specific noise.

Therefore, the Factor Analysis dashboard correctly reveals that `Sensor_4` contributes almost no useful structural information despite PCA assigning importance to its large variance. This demonstrates why PCA alone can be misleading for industrial anomaly detection and why variance decomposition through Factor Analysis provides a more reliable diagnostic framework.

### Question 2: Decoupling Structural Loading via Rotation (FA Subplot 1 vs. PCA Eigenvectors)

* **Context:** Look at the **Structural Loadings Matrix** heatmap in the top-left of the FA dashboard. You can see that `Sensor_1` and `Sensor_2` load heavily on `Factor 1`, while `Sensor_3` loads exclusively on `Factor 2`.
* **Core Problem:**
1. Traditional PCA forces its principal axes to follow a strict mathematical hierarchy ($\lambda_1 > \lambda_2 > \dots$) where every component tries to sweep up as much mixed variance as possible across all sensors. Explain how the **Varimax rotation** in Factor Analysis relaxes this rule to create "simple structure" (clean, isolated groupings).
2. From a plant operator's standpoint, why is the rotated FA heatmap much easier to troubleshoot during a structural failure than a raw, unrotated PCA eigenvector matrix?





## Interpretation of the Rotated Factor Analysis Loadings Matrix

### 1. How Varimax Rotation Creates a Simple Structure

In Principal Component Analysis (PCA), the principal components are obtained by solving the eigenvalue problem

$$
\Sigma = P\Lambda P^T,
$$

where

$$
\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_m.
$$

The first principal component is forced to capture the maximum possible variance, the second captures the largest remaining variance, and so forth.

As a consequence, PCA components often contain contributions from many sensors simultaneously. A typical PCA loading vector may appear as

$$
PC_1 =
0.48S_1 +
0.52S_2 +
0.44S_3 +
0.55S_4.
$$

Although mathematically optimal for variance capture, such a component mixes several physical sources together.

---

Factor Analysis introduces a loading matrix

$$
\Lambda
=
[\lambda_{jr}]
$$

and then applies an orthogonal rotation matrix

$$
T
$$

to obtain

$$
\Lambda_{\text{rotated}}
=
\Lambda T.
$$

Varimax rotation chooses \(T\) so that each factor contains:

- A few very large loadings,
- Many near-zero loadings.

This process is called **simple structure**.

Instead of every factor sharing variance across all sensors, the rotation encourages clean separation of sensor groups.

For example, after Varimax rotation the loading matrix may resemble

$$
\begin{array}{c|cc}
 & F_1 & F_2\\
\hline
S_1 & 0.91 & 0.05\\
S_2 & 0.88 & 0.02\\
S_3 & 0.03 & 0.93\\
S_4 & 0.07 & 0.10
\end{array}
$$

In this structure:

- `Sensor_1` and `Sensor_2` belong primarily to Factor 1.
- `Sensor_3` belongs primarily to Factor 2.
- `Sensor_4` contributes weakly to both factors.

The factors become physically interpretable rather than purely mathematical variance directions.

---

### 2. Why the Rotated FA Heatmap Is Easier to Troubleshoot

From a plant operator's perspective, the objective is not merely to maximize variance explained. The objective is to identify which physical subsystem is responsible for abnormal behavior.

A raw PCA eigenvector matrix often contains mixed loadings across many sensors:

$$
PC_1
=
0.48S_1
+
0.52S_2
+
0.44S_3
+
0.55S_4.
$$

If an anomaly appears in \(PC_1\), it is difficult to determine which sensor group or physical mechanism caused the event because all sensors contribute simultaneously.

The operator must interpret a mathematical combination of multiple signals rather than a distinct physical subsystem.

---

In contrast, the rotated FA heatmap produces localized factor groupings.

For example:

- Factor 1 may represent a structural load pathway monitored by `Sensor_1` and `Sensor_2`.
- Factor 2 may represent a different subsystem monitored primarily by `Sensor_3`.

If an abnormal factor score is observed:

- A rise in Factor 1 immediately directs attention toward the subsystem associated with `Sensor_1` and `Sensor_2`.
- A rise in Factor 2 immediately points toward the subsystem associated with `Sensor_3`.

This greatly simplifies fault isolation.

---

### Engineering Significance

The PCA loading matrix answers:

> "Which mathematical directions explain the most variance?"

The rotated Factor Analysis loading matrix answers:

> "Which groups of sensors are responding to the same underlying physical mechanism?"

Therefore, during structural failures, equipment degradation, or process disturbances, the Varimax-rotated Factor Analysis heatmap is substantially easier to interpret because it produces clean, physically meaningful sensor clusters rather than mixed variance directions. This makes troubleshooting, root-cause analysis, and maintenance decisions much faster and more reliable.

### Question 3: Determining Subspace Truncation ($k$) using $T^2$ and $Q$ Profiles

* **Context:** Look at the right side of the PCA dashboard, showing **Mean Hotelling's $T^2$** and **Mean $Q$ Statistic (SPE)** plotted against the truncation cutoff $k$.
* **Core Problem:**
1. Describe the exact behavior of the Mean $Q$ Statistic curve as the subspace cutoff transitions from $k=1$ to $k=2$, and then from $k=2$ to $k=3$.
2. Why does a sharp drop followed by a distinct "flat elbow" in the $Q$ statistic identify **$k=2$** as the true hidden physical dimensionality of this asset?
3. If you incorrectly choose a truncation cutoff of $k=3$ for your monitoring system, what type of data are you forcing into your "clean" principal subspace?

## Interpretation of the Mean Q Statistic (SPE) Curve

### 1. Behavior of the Mean Q Statistic as \(k\) Increases

The Q Statistic (Squared Prediction Error) measures the residual energy remaining outside the retained PCA subspace.

Mathematically,

$$
Q_i
=
\sum_{j=k+1}^{m}
z_{i,j}^{2},
$$

where the retained components are excluded and only the discarded residual coordinates contribute.

---

### Transition from \(k=1\) to \(k=2\)

When only one principal component is retained (\(k=1\)), a substantial portion of the system variance remains in the residual space.

The Mean Q Statistic is therefore relatively large.

After increasing the cutoff to

$$
k=2,
$$

the second major principal component is incorporated into the retained subspace.

Since the second component contains a large amount of structured system variance, a significant amount of residual energy is removed.

Consequently,

$$
\boxed{
\text{Mean }Q
\text{ decreases sharply from }
k=1
\text{ to }
k=2
}
$$

This sharp drop indicates that the second component captures important physical information rather than noise.

---

### Transition from \(k=2\) to \(k=3\)

When the cutoff is increased further to

$$
k=3,
$$

the Mean Q Statistic decreases only slightly.

The residual curve begins to flatten because most of the meaningful system variance has already been captured by the first two principal components.

Therefore,

$$
\boxed{
\text{Mean }Q
\text{ shows only a small decrease from }
k=2
\text{ to }
k=3
}
$$

This behavior produces a characteristic elbow in the curve.

---

### 2. Why the Elbow Identifies \(k=2\) as the True Physical Dimensionality

A PCA model seeks the smallest subspace that captures the dominant physical structure of the system.

If the first two principal components explain nearly all meaningful covariance relationships, then

$$
\mathbf{x}
\approx
\mathbf{P}_2\mathbf{z}_2.
$$

Adding additional components after \(k=2\) contributes little new information.

The Q curve therefore exhibits:

- A large reduction when moving from \(k=1\) to \(k=2\),
- A much smaller reduction when moving from \(k=2\) to \(k=3\).

This creates an elbow shape.

Graphically:

```text
Q
│
│\
│ \
│  \
│   \____
│        \__
└────────────── k
    1    2   3
```

The elbow occurs at the point where the major structured variance has been captured and only weak residual variation remains.

Therefore,

$$
\boxed{
k=2
}
$$

is interpreted as the intrinsic dimensionality of the underlying physical system.

In other words, the asset behaves as if it is governed primarily by two independent physical mechanisms.

---

### 3. Consequences of Choosing \(k=3\)

When the cutoff is increased beyond the true system dimensionality,

$$
k > 2,
$$

the PCA model begins absorbing residual variation that should remain outside the principal subspace.

This additional variation often consists of:

- Measurement noise,
- Sensor-specific disturbances,
- Environmental fluctuations,
- Numerical uncertainty,
- Instrument drift.

Consequently, selecting

$$
k=3
$$

forces part of the residual noise space into the retained PCA model.

Mathematically,

$$
\mathbf{x}
=
\mathbf{P}_2\mathbf{z}_2
+
\mathbf{P}_{\text{noise}}
\mathbf{z}_{\text{noise}}
$$

and the third component begins representing portions of

$$
\mathbf{P}_{\text{noise}}
\mathbf{z}_{\text{noise}}.
$$

rather than meaningful structural behavior.

---

### Engineering Interpretation

Choosing \(k=3\) causes the monitoring system to treat some noise as if it were legitimate system dynamics.

This can produce several undesirable effects:

1. Reduced sensitivity of the Q Statistic to structural faults.
2. Increased model complexity without meaningful information gain.
3. Greater risk of masking emerging damage signatures.
4. Increased likelihood of overfitting the baseline data.

Therefore, the elbow in the Mean Q Statistic curve provides strong evidence that

$$
\boxed{
k=2
}
$$

is the most appropriate truncation boundary because it captures the true physical structure while leaving sensor noise and unexplained disturbances in the residual monitoring space.

### Question 4: Operational Trade-offs in System Health Monitoring

* **Context:** Imagine you are deploying a real-time predictive maintenance pipeline for a fleet of identical industrial assets based on these exact models.
* **Core Problem:** Compare the operational trade-offs of these two approaches:
1. **The PCA Strategy:** Tracking the main principal subspace using Hotelling's $T^2$ alongside a residual noise floor monitor using the $Q$ statistic.
2. **The FA Strategy:** Monitoring the rotated Latent Factor Scores directly.


* Which strategy is more robust against a single sensor losing calibration or experiencing an electrical short? Justify your choice using the **Sensor Uniqueness Noise Floor ($\varphi^2$)** metrics.

## Comparison of PCA-Based and Factor Analysis-Based Predictive Maintenance Strategies

### 1. PCA Monitoring Strategy

The PCA approach decomposes the observation space into:

- A retained principal subspace monitored using Hotelling's \(T^2\),
- A residual subspace monitored using the \(Q\) statistic (SPE).

The monitoring statistics are

$$
T^2
=
\mathbf{z}_k^T
\mathbf{\Lambda}_k^{-1}
\mathbf{z}_k
$$

and

$$
Q
=
\|\mathbf{e}\|^2.
$$

The \(T^2\) statistic measures how far an observation moves within the learned principal operating space, while the \(Q\) statistic measures energy that falls outside the PCA model.

This strategy is highly effective for detecting:

- Changes in operating conditions,
- Global process shifts,
- Environmental loading changes,
- Multivariate deviations from the baseline operating region.

Because PCA maximizes total variance, it captures dominant system behavior using a compact set of principal components.

---

### 2. Factor Analysis Monitoring Strategy

Factor Analysis models the system using latent physical factors:

$$
\mathbf{x}
=
\boldsymbol{\Lambda}\mathbf{f}
+
\boldsymbol{\epsilon}.
$$

The monitoring variables become the rotated factor scores

$$
\mathbf{f}
=
\boldsymbol{\Lambda}^T
\mathbf{R}^{-1}
\mathbf{z}.
$$

Each factor score represents a physically interpretable latent mechanism extracted from the sensor network.

Instead of monitoring variance directions, the FA framework monitors:

- Structural load pathways,
- Common physical behaviors,
- Latent system states.

The rotation process produces a simple structure where groups of sensors are associated with specific factors, making fault isolation significantly easier.

---

## Robustness to Sensor Calibration Loss or Electrical Shorts

Suppose a single sensor experiences:

- Calibration drift,
- Electrical interference,
- Wiring faults,
- Amplifier saturation,
- Complete sensor degradation.

Such failures generate variance that is localized to one channel rather than shared across the entire sensor network.

Factor Analysis explicitly separates total variance into

$$
1
=
h_j^2
+
\varphi_j^2,
$$

where:

- \(h_j^2\) = communality (shared structural variance),
- \(\varphi_j^2\) = uniqueness (sensor-specific variance).

A malfunctioning sensor typically exhibits

$$
\varphi_j^2 \rightarrow 1.
$$

This means that most of its variance is isolated from the common latent factors.

Consequently, Factor Analysis naturally identifies problematic sensors through elevated uniqueness values.

For example, if

$$
\varphi_4^2 = 0.98,
$$

then approximately \(98\%\) of the variance observed by Sensor 4 is sensor-specific and contributes very little information about the actual structural state.

The uniqueness profile therefore acts as a built-in sensor health indicator.

---

## Why PCA Is More Vulnerable

PCA does not distinguish between:

- Shared structural variance,
- Sensor-specific noise.

Instead, it maximizes total variance:

$$
\Sigma
=
P\Lambda P^T.
$$

A faulty sensor with large variance can therefore increase the covariance matrix and artificially inflate one or more eigenvalues.

As a result:

- Noise may be absorbed into the principal subspace.
- Hotelling's \(T^2\) thresholds may become distorted.
- Principal components may partially represent sensor faults rather than true system dynamics.

The PCA model may incorrectly interpret sensor degradation as meaningful process variation.

---

## Engineering Conclusion

The PCA strategy provides an excellent framework for detecting global operational changes through the combined use of:

- Hotelling's \(T^2\),
- Residual \(Q\) statistics.

However, PCA treats all variance equally and cannot inherently distinguish physical behavior from localized sensor faults.

Factor Analysis is generally more robust against calibration loss or electrical shorts because the uniqueness metric

$$
\varphi_j^2
$$

explicitly quantifies sensor-specific variance.

A faulty sensor naturally produces an elevated uniqueness value while contributing little to the common latent factors.

Therefore,

$$
\boxed{
\text{Factor Analysis is more robust to single-sensor failures}
}
$$

because the Sensor Uniqueness Noise Floor directly identifies localized measurement problems before they contaminate the latent structural state estimates.

In practical predictive maintenance systems, PCA is highly effective for detecting overall process anomalies, while Factor Analysis provides superior robustness and fault isolation when individual sensors degrade or fail. The most reliable industrial monitoring architectures often combine both approaches: PCA for global anomaly detection and Factor Analysis for sensor-level diagnostic interpretation.